### PR TITLE
feat: NVIDIA-style Hierarchical Gradient Checkpointing

### LINKED ISSUE #189: Optimization: Gradient checkpointing for deep processor layers
## Detailed Description
Following the merge of **Efficient Sequential Batching** (PR #187), the memory bottleneck related to graph edges has been resolved.
However, the model still consumes significant GPU memory storing intermediate activations for backpropagation, particularly in the deep GraphProcessor typically 9 layers.i assure this activation memory scales linearly 

This issue proposes implementing gradient checkpointing to trade a small amount of compute re calculating forward passes during backward pass for substantial memory savings

## Context
This optimization complements the recent batching fixes and brings the architecture closer to implementations like **NVIDIA's GraphCast** https://github.com/NVIDIA/physicsnemo/blob/main/physicsnemo/models/graphcast/graph_cast_net.py
till now i can't run 1degree in my system(4gb gpu).after this it is expected to significantly improve support for low vram systems

## Possible Implementation
Add a use_checkpointing boolean flag to GraphWeatherForecaster and Processor
in processor.forward wrap the iterative block updates in the checkpoint function

### PR DESCRIPTION
**Pull request by Sidharth1743**
# Pull Request

## Description

Implemented NVIDIA graphcast Hierarchical Gradient Checkpointing for the GraphProcessor and related layers, significantly reducing GPU memory usage during training.

Summary of Changes:
Hierarchical Checkpointing: Implemented a new GraphCast wrapper class that allows fine-grained control over checkpointing (Model, Encoder, Processor, Decoder levels).
Core Optimization: Modified graph_net_block.py to use torch.utils.checkpoint with use_reentrant=False and preserve_rng_state=False. This matches NVIDIA's PhysicsNeMo implementation for maximum memory efficiency
New API: Added a configuration interface (GraphCastConfig) with pre-set strategies (e.g., balanced, full, processor_only)

Motivation: Anyone can see future weather with ow vram gpu

Fixes #189 

## How Has This Been Tested?

I have added a new test suite and updated the benchmarking script
1. Unit Tests: Ran pytest tests/models/test_gradient_checkpointing.py which verifies:
Output Equivalence: Confirmed that model(x) produces the exact same output whether checkpointing is True or False.
Gradient Flow: Verified that gradients propagate correctly through the checkpointed layers.
Hierarchical Control: Verified that flags correctly toggle checkpointing for specific sub-modules.

2. Memory Benchmarks: Ran python scripts/benchmark_memory.py on an NVIDIA RTX 3050 (4GB VRAM).
Results:
5.0° Grid (Batch 1): Memory dropped from 3012 MB (Baseline) to 856 MB (Checkpointed). (71.6% Reduction)
2.5° Grid (Batch 1-4): Previously failed with OOM. Now runs successfully (Peak mem: ~1.2GB - 2.6GB).
[x] Yes

If your changes affect data processing, have you plotted any changes? i.e. have you done a quick sanity check?

[x] Yes (Verified numerical equivalence of model outputs)

## Checklist:

- [ x] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [ x] I have performed a self-review of my own code
- [ x] I have made corresponding changes to the documentation
- [ x] I have added tests that prove my fix is effective or that my feature works
- [ x] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by Sidharth1743** [comment]
sorry @jacobbieker i don't have the permissions to use the 'Request Review' button on this new PR yet(no gear icon visible)
ready for review!

File: graph_weather/models/layers/processor.py [hunk 1]
```diff
@@ -59,10 +62,32 @@ def __init__(
...
         )
         if self.use_thermalizer:
             self.thermalizer = ThermalizerLayer(input_dim)
 
+    def set_checkpoint_segments(self, checkpoint_segments: int):
+        """Sets checkpoint segments for the processor.
+
+        This matches NVIDIA's API for controlling processor checkpointing.
+
+        Parameters
```
Comment (jacobbieker) [inline graph_weather/models/layers/processor.py]: Can you make this match the other docstring styles? Just want to keep it consistent

File: graph_weather/models/graphcast.py [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+from graph_weather.models.layers.processor import Processor
+
+
+class GraphCast(torch.nn.Module):
+    """GraphCast model with hierarchical gradient checkpointing.
+
+    This model combines Encoder, Processor, and Decoder with NVIDIA-style
+    hierarchical checkpointing controls for flexible memory-compute tradeoffs.
+
+    Parameters
```
Comment (jacobbieker) [inline graph_weather/models/graphcast.py]: Please adjust this to match the docstyle of the other docstrings in this repo

File: graph_weather/models/graphcast.py [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+        Only effective when set_checkpoint_model(False).
+
+        Parameters
+        ----------
+        checkpoint_flag : bool
+            If True, checkpoint decoder section.
+        """
+        self._checkpoint_decoder = checkpoint_flag
+
+    def _set_layer_checkpointing(self, flag: bool):
```
Comment (jacobbieker) [inline graph_weather/models/graphcast.py]: Remove this

File: scripts/benchmark_gradient_checkpointing.py [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+        else:
+            raise
+
+    finally:
+        # Cleanup
+        del model
+        tracker.reset()
+
+
+def print_results_table(results: List[BenchmarkResult], grid_resolution: float):
```
Comment (jacobbieker) [inline scripts/benchmark_gradient_checkpointing.py]: Please remove all these print statements, just have the script output to a file. Also, could this be combined with the other benchmarking script? Or refactored for similar outupts? Would prefer not to just keep adding large, very similar, benchmarking scripts

File: tests/models/test_gradient_checkpointing.py [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+            x = processor_with_cp(
+                x, edge_idx, edge_attr, batch_size=batch_size, efficient_batching=True
+            )
+            out_with_cp = decoder_with_cp(x, features, batch_size)
+
+        # Verify outputs are identical
+        assert torch.allclose(out_no_cp, out_with_cp, atol=1e-5)
+
+
+class TestGraphCastModel:
```
Comment (jacobbieker) [inline tests/models/test_gradient_checkpointing.py]: I prefer not having classes for the test suite with pytest, so please remove the classes here and have it just be the test functions

File: tests/models/test_gradient_checkpointing.py [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+    """Test that preserve_rng_state=False is correctly used."""
+
+    def test_checkpoint_uses_correct_flags(self):
+        """Test that checkpointing uses use_reentrant=False and preserve_rng_state=False."""
+        # This is tested implicitly by the other tests working correctly
+        # The flags are hardcoded in the implementation
+        pass
+
+
+if __name__ == "__main__":
```
Comment (jacobbieker) [inline tests/models/test_gradient_checkpointing.py]: Test files should not have a block like this, so remove it. The tests should run with `pytest .` with all the other tests

**Review by jacobbieker — CHANGES_REQUESTED** [review]
Overall thanks for adding this. The docstyle needs to be fixed to match the style used for the rest of the repository. Additionally, I would rather not have a continously increasing number of very similar large benchmarking scripts. I would suggest refactoring the current one to also be able to run this benchmark, since both are benchmarking memory savings with different changes. Finally, the tests should be reorganized.

File: graph_weather/models/layers/processor.py (hunk 1 above)
Comment (Sidharth1743) [reply graph_weather/models/layers/processor.py]: yes did it

File: graph_weather/models/graphcast.py (hunk 2 above)
Comment (Sidharth1743) [reply graph_weather/models/graphcast.py]: sure

File: scripts/benchmark_gradient_checkpointing.py (hunk 4 above)
Comment (Sidharth1743) [reply scripts/benchmark_gradient_checkpointing.py]: sure i thought the we can have two benchamrk for training and inference but it will be good in single file as belongs to memory
modifying

File: tests/models/test_gradient_checkpointing.py (hunk 5 above)
Comment (Sidharth1743) [reply tests/models/test_gradient_checkpointing.py]: done

File: tests/models/test_gradient_checkpointing.py (hunk 6 above)
Comment (Sidharth1743) [reply tests/models/test_gradient_checkpointing.py]: sorry my mistake
will make the changes shortly

File: graph_weather/models/graphcast/model.py
Comment (jacobbieker) [inline graph_weather/models/graphcast/model.py:1]: This should be under `models/graphcast/model.py` please

**Review by jacobbieker — APPROVED** [review]
Looks nice, thanks for this. Just one minor moving around then we can merge

File: graph_weather/models/graphcast/model.py
Comment (Sidharth1743) [reply graph_weather/models/graphcast/model.py:1]: no problem i will do this

**Review by jacobbieker — APPROVED** [review]
LGTM! Thanks for this.
